File: ios/SwiftKotlination/SwiftKotlination/Sources/Features/TopStories/ViewModel/TopStoriesViewModel.swift
import UIKit.UIImage

final class TopStoriesViewModel {

    typealias Factory = ImageFactory & TopStoriesFactory
    private let factory: Factory

    private lazy var topStoriesManager = factory.makeTopStoriesManager()
    private lazy var imageManager = factory.makeImageManager()

    init(factory: Factory) {
        self.factory = factory
    }

    var stories: [Story] = []
    var images: [String: UIImage] = [:]

    @discardableResult
    func stories(_ observer: @escaping Observer<[Story]>) -> Disposable {
        return topStoriesManager
            .stories { [weak self] result in
                switch result {
                case .success(let stories):
                    self?.stories = stories

                case .failure:
                    break
                }
                observer(result)
        }
    }

    func refresh() {
        topStoriesManager.fetchStories()
    }

    @discardableResult
    func image(with url: String, _ observer: @escaping Observer<UIImage>) -> Disposable? {
        guard let image = images[url] else {
            return imageManager
                .image(with: url) { [weak self] result in
                    switch result {
                    case .success(let image):
                        self?.images[url] = image

                    case .failure:
                        break
                    }
                    observer(result)
            }
        }

        observer(.success(image))
        return nil
    }
}

File: android/SwiftKotlination/features/topstories/src/main/kotlin/fr/jhandguy/topstories/viewmodel/TopStoriesViewModel.kt
package fr.jhandguy.topstories.viewmodel

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import fr.jhandguy.network.model.observer.Disposable
import fr.jhandguy.network.model.observer.Observer
import fr.jhandguy.network.model.observer.Result
import fr.jhandguy.story.model.Story
import fr.jhandguy.topstories.factory.TopStoriesFactory

class TopStoriesViewModel(factory: TopStoriesFactory) {
    private val topStoriesManager = factory.makeTopStoriesManager()
    private val imageManager = factory.makeImageManager()

    var stories: List<Story> = emptyList()
    var images: MutableMap<String, Bitmap> = HashMap()

    suspend fun topStories(observer: Observer<List<Story>>) = topStoriesManager.topStories { result ->
        when (result) {
            is Result.Success -> {
                stories = result.data.results
                observer(Result.Success(stories))
            }
            is Result.Failure -> observer(result)
        }
    }

    suspend fun refresh() = topStoriesManager.fetchStories()

    suspend fun image(url: String, observer: Observer<Bitmap>): Disposable? = images[url]?.let { image ->
        observer(Result.Success(image))
        return null
    } ?: imageManager.image(url) { result ->
        when (result) {
            is Result.Success -> {
                val image = BitmapFactory.decodeByteArray(result.data, 0, result.data.size)
                images[url] = image
                observer(Result.Success(image))
            }
            is Result.Failure -> observer(result)
        }
    }
}
